File: 12b/main.py
from collections import Counter


with open('input.txt', 'rt') as f:
    raw = [l.strip().split('-') for l in f.readlines()]
    conns = {}
    for a, b in raw:
        if a in conns:
            conns[a].append(b)
        else:
            conns[a] = [b]
        if b in conns:
            conns[b].append(a)
        else:
            conns[b] = [a]

path = ['start']
cant_visit = ['start']

def is_valid(path: list[str]):
    relkeys = [p for p in path if p.islower() and p != 'start']
    counter = Counter(relkeys)
    okay = True
    for v in counter.values():
        if v >= 2:
            if okay:
                okay = False
            else:
                return False
    return True

viable = []
def paths(path: list[str], cant_visit: list[str], has_double: bool) -> list[list[str]]:
    if path[-1] == 'end':
        viable.append(path)
        return
    accessible = [
        n for n in conns[path[-1]]
        if n not in cant_visit
    ]
    for a in accessible:
        new_has_double = has_double
        if a in path and a.islower():
            new_cant_visit = [n for n in path if n.islower()] + ['start']
            new_has_double = True
        elif has_double and a.islower():
            new_cant_visit = cant_visit + [a]
        else:
            new_cant_visit = cant_visit
        paths(path + [a], new_cant_visit, new_has_double)

paths(path, cant_visit, False)

print(len(viable))

File: 12b/ruby/main.rb
require "Set"

FILENAME = "input.txt"

lines = File.open(FILENAME).readlines.map(&:chomp)
Paths = {}

lines.each do |line|
  a, b = line.split("-")
  # Add links from a to b and b to a
  Paths[a] = [] unless Paths.has_key? a
  Paths[a] << b
  Paths[b] = [] unless Paths.has_key? b
  Paths[b] << a
end

class String
  def is_lower?
    self == self.downcase
  end
  def is_upper?
    self == self.upcase
  end
end

def get_paths(current_path, seen, can_revisit=true)
  current_cave = current_path[-1]
  if current_cave == "end"
    return [current_path]
  end
  # Rule out caves that are small and already-seen
  available_caves = Paths[current_cave].reject{|cave| cave == "start"}.select do |cave|
    cave.is_upper? or can_revisit or (cave.is_lower? and not seen.include?(cave))
  end
  paths = available_caves.map do |cave|
    if cave.is_lower?
      if seen.include?(cave)
        new_path = current_path + [cave]
        get_paths(new_path, seen, false)
      else
        new_seen = seen + Set[cave]
        new_path = current_path + [cave]
        get_paths(new_path, new_seen, can_revisit)
      end
    else
      get_paths(current_path + [cave], seen, can_revisit)
    end
  end
  paths.flatten 1
end

paths = get_paths(["start"], Set["start"]).uniq
p paths.length
